snakemake/uf_cells_spleen_0510/src/generate_cell_sets.py

```diff
@@ -0,0 +1,170 @@
+    tree = init_tree()
+
+    leiden_clusters_children = []
+    for cluster_name, cluster_df in df.groupby("leiden"):
+        leiden_clusters_children.append({
+            "name": cluster_name,
+            "set": cluster_df[COLUMNS.CELL_ID.value].unique().tolist()
+        })
+
+    tree["tree"].append({
+        "name": "Leiden Clustering",
+        "children": leiden_clusters_children
+    })
+
+    cell_type_annotation_children = []
+    for cell_type, cell_type_df in df.groupby(COLUMNS.ANNOTATION.value):
+        cell_type_annotation_children.append({
+            "name": cell_type,
+            "set": cell_type_df[COLUMNS.CELL_ID.value].unique().tolist()
+        })
+
+    tree["tree"].append({
+        "name": "Cell Type Annotations (flat)",
+        "children": cell_type_annotation_children
+    })
+    return tree
+
+
+def generate_hierarchical_cell_sets(df, cl_obo_file):
+    tree = init_tree()
+
+    # Load the cell ontology DAG
+    graph, id_to_name, name_to_id = load_cl_obo_graph(cl_obo_file)
+
+    ancestors_and_sets = []
+
+    for cell_type, cell_type_df in df.groupby(COLUMNS.ANNOTATION.value):
+
+        try:
+            node_id = name_to_id[cell_type]
+        except KeyError:
+            print((
+                f"ERROR: annotation '{cell_type}' does "
+                "not match any node in the cell ontology."
+            ))
+
+        # Get ancestors of the cell type
+        # (counterintuitive that the function is called descendants)
+        ancestor_term_set = networkx.descendants(graph, node_id)
+
+        # Make sure the cell type has an ancestor
+        # with the 'cell' root ID
+        assert(CL_ROOT_ID in ancestor_term_set)
+
+        # Initialize the current node ID to the cell type of interest
+        curr_node_id = node_id
+
+        # Construct a list of ancestors, inclusive.
+        # [node_id, parent_id, grandparent_id, ...]
+        ancestors = [curr_node_id]
+
+        # Get the parents of the current node.
+        curr_node_parents = list(graph.out_edges(curr_node_id, keys=True))
+
+        # Warn if the current node has multiple parents.
+        if len(curr_node_parents) > 1:
+            print((
+                f"WARN: {curr_node_id} has "
+                f"{len(curr_node_parents)} parents."
+            ))
+
+        # Select the first (hopefully only) parent node.
+        _, curr_parent_id, relationship = curr_node_parents[0]
+        assert(relationship == "is_a")
+
+        while curr_node_id != CL_ROOT_ID:
+            # Get the parents of the current node.
+            curr_node_parents = list(graph.out_edges(curr_node_id, keys=True))
+
+            # Warn if the current node has multiple parents.
+            if len(curr_node_parents) > 1:
+                print((
+                    f"WARN: {curr_node_id} has "
+                    f"{len(curr_node_parents)} parents."
+                ))
+
+            # Select the first (hopefully only) parent node.
```

Comment:
Have you seen >1 parent?  Could that really break things?

Reply:
Yes there are a lot with 2 or 3 parents. It does not break things because I just naïvely choose the first parent. But not sure if this is a way to identify the "correct" or "most important" parent when this occurs.